add child device tutorial (#1642)
thin-edge/thin-edge.io: docs/src/tutorials/child-device-config-mngmnt.md

> # Child devices: configuration management

Reviewer: ```suggestion
# Configuration management on child devices
```

After following this tutorial, you will know how to manage various configuration files  regarding child devices. For this tutorial a Raspberry Pi is used. In a few steps will be explained how get a configuration snapshot from a child device and also sending configuration information to a child device.

## Introduction
Lets first start with the definition of child devices.

According to this [document](https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/architecture/domain-model.md), thin-edge.io facilitates IoT functionality to the device it is running on, as well as to devices that are connected to that device.
  * the device thin-edge is running on is referred as the main-device
    * thin-edge on the main-device establishes and manages all communication to the cloud
  * all devices connected to the main-device are referred as external child-devices
  * each external child-device can be represented in the cloud with its individual device twin**
    * a unique child-id makes the association between each external child-device and it's device twin**
  * all telemetry data and device management functionality can appear in the context of the

>     external child-device's individual device twin, or the main-device's device twin

Author: I changed it accordingly, was unclear

Since child device agents typically run on an external device and not on the thin-edge.io device itself, the MQTT and HTTP API's of thin-edge.io need to be accessed over the network using its IP address, which is configured using the thin-edge.io configuration settings `mqtt.external.bind_address` or `mqtt.bind_address`. The MQTT APIs are exposed via port 1883 and the HTTP API's are exposed via port 8000. In rare cases, where the child device agent is installed alongside thin-edge.io on the same device, these API's can be accessed via a local IP or even 127.0.0.1.

In this tutorial 127.0.0.1. is used.

##  Prerequisite

To follow this tutorial, you only need the following:

- A [Cumulocity IoT](https://www.softwareag.cloud/site/product/cumulocity-iot.html) tenant.

> - Installed the latest version of thin-edge.io. To this, you can follow [this tutorial](https://thin-edge.github.io/thin-edge.io/html/tutorials/getting-started.html#Step-6-Manage-configuration-files), which will also give you some deeper insights in the various capabilities of thin-edge.io, like configuration management.

Reviewer: Install where? On the Raspberry Pi? In that case you need to put the Raspberry Pi as req. 2 and this as req. 3.

- A Raspberry Pi (3 or 4) with Raspian installed, configured and connected to the Cumulocity tenant. For other boards and OS'es have a look [here](https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/supported-platforms.md)

> - Updated device: `$ sudo apt-get update && sudo apt-get upgrade`

Author: Are you pointing to the last line about updating?

## Steps

This tutorial is divided into 3 main steps.

[Step 1 Startup / Bootstrap of the child device](#Startup-/-Bootstrap-of-the-child-device).

In this step a child device with its supported configuration list will be created and will be sent to thin-edge.io during the startup/bootstrap phase of the child device agent.

[Step 2 Handle config snapshot request from thin-edge.io](#handle-config-snapshot-request-from-thin-edge.io)

In this step configuration files from a child device will be requested to make them visible in Cumulocity IoT.

[Step 3 Handle config update request from thin-edge.io](#handle-config-update-request-from-thin-edge.io)

> In this configuration files from a child device will be updated.

Reviewer: ```suggestion
In this configuration files from a child device will be updated with the updated configuration file pushed from Cumulocity IoT.
```

### Step 1 Startup / Bootstrap of the child device

> In this step a child device with its supported configuration list will be created and will be sent to thin-edge.io during the startup/bootstrap phase of the child device agent.

Reviewer: ```suggestion
In this step, a child device with its supported configuration list will be created and will be sent to thin-edge.io.
```
We're already in the section titled "Startup / Bootstrap of the child device". So, that second part of the sentence looked redundant.

Therefore some steps have to be taken

> 1. Create a `c8y-configuration-plugin.toml` this file consists of the supported configuration list i.e. a list of configuration files.

Reviewer: ```suggestion
1. Create a `c8y-configuration-plugin.toml` file that contains the supported configuration list i.e. a list of configuration files.
```

> Note: this is a separate file and not the already exsisting `c8y-configuration-plugin.toml` file which is used for configuration management of thin-edge.io.

Reviewer: ```suggestion
Note: This is not the `c8y-configuration-plugin.toml` file which is used for configuration management of thin-edge.io main device.
This is a separate file, required for each child device, with its supported configuration list.```

> The child device agent needs to capture the list of configuration files that needs be managed from the cloud in a `c8y-configuration-plugin.toml` file in the same format as specified in the [configuration management documentation](https://thin-edge.github.io/thin-edge.io/html/howto-guides/025_config_management_plugin.html) as follows:

Reviewer: Could you combine this line with the bullet point.1 above, to avoid redundancy.

Author: done

```
files = [
    { path = '/path/to/some/config', type = 'config1'},
]
```
* `path` is the full path to the configuration file on the child device file system.
* `type` is a unique alias for each file entry which will be used to represent that file in Cumulocity IoT.

Example:

```
files = [
    { path = '/home/pi/config/config1', type = 'config1'},
    { path = '/home/pi/config/config2', type = 'config2'},
]
```

> Note: these referenced files including the path should be created.

Reviewer: ```suggestion

Note: Update these paths with some realistic paths on your device, or create these files.
```

2. Upload this file to thin-edge.io via HTTP: The child device agent needs to upload this file to thin-edge.io with an HTTP PUT request to the URL:

`
http://{tedge-ip}:8000/tedge/file-transfer/{child-id}/c8y-configuration-plugin
`

* `{tedge-ip}` is the IP of the thin-edge.io device which is configured as `mqtt.external.bind_address` or `mqtt.bind_address` or `127.0.0.1` if neither is configured.
* `{child-id}` is the child-device-id

> example: `curl -X PUT --data-binary @/home/pi/config/c8y-configuration-plugin.toml http://127.0.0.1:8000/tedge/file-transfer/child1/c8y-configuration-plugin

Reviewer: ```suggestion
Example:
```console
curl -X PUT --data-binary @/home/pi/config/c8y-configuration-plugin.toml http://127.0.0.1:8000/tedge/file-transfer/child1/c8y-configuration-plugin
```
```

`

3. Notify thin-edge.io about the upload via MQTT

Once the upload is complete, the agent should notify thin-edge.io about the upload by sending the following MQTT message:

Topic:

`tedge/{child-d}/commands/res/config_snapshot`

Payload:

`{ "type": "c8y-configuration-plugin”, "path": ”/child/local/fs/path”}`

Example:

> Note: For using the `mosquito-pub` command (and later on in this tutorial also the `mosquitto_sub` command), you need to install the  mosquitto clients via `sudo apt-get install mosquitto-clients`

Reviewer: ```suggestion
Note: For using the `mosquito-pub` command (and later on in this tutorial also the `mosquitto_sub` command), you need to install the  mosquitto clients via `sudo apt-get install mosquitto-clients`
```

Move this note below the command and also highlight that any MQTT library/client could be used instead of the `mosquitto` ones.

Author: changed this note

> `mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_snapshot" -m '{"status": null,  "path": "", "type":"c8y-configuration-plugin", "reason": null}'

Reviewer: ```suggestion
`mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_snapshot" -m '{"path": "", "type":"c8y-configuration-plugin"}'
```

`

> 

Reviewer: You could show the configuration list appearing in the Cumulocity UI here.

Author: added the configuration list

> ### Step 2 Handle config snapshot request from thin-edge.io

Reviewer: ```suggestion
### Step 2 Handle config snapshot request
```

In this step configuration files from a child device will be requested to make them visible in Cumulocity IoT.

> Handling config snapshot requests from thin-edge.io is a 4-step process:

Reviewer: ```suggestion
Handling config snapshot request is a 6-step process:
```
As the below steps involve things done from the Cloud as well as the child device, and there's nothing explicitly done on the thin-edge device.

> 1. Subscribe to, and receive config snapshot requests via MQTT: `mosquitto_sub -h localhost -t "#"`

Reviewer: ```suggestion
1. Subscribe to, and receive config snapshot requests via MQTT: `mosquitto_sub -h 127.0.0.1 -t "tedge/{child-id}/commands/req/config_snapshot"`
```

Keep the usage of `127.0.0.1` or `localhost` consistent across all the commands.

> 2. In Cumulocity IoT, go to `Device Management` -> `All Devices` -> `Devices`, then click on your device. Then click on  `Child devices`  then click on the in *Step 1 Startup / Bootstrap of the child device* defined child device (child1) and then click on `Configuration`.

Reviewer: ```suggestion
2. In Cumulocity IoT, go to `Device Management` -> `All Devices` -> `Devices`, then click on your device. Then click on `Child devices`, then click on the child device (child1) created in *Step 1: Startup / Bootstrap of the child device* and then click on `Configuration`.
```
An image here with the relevant buttons highlighted would be nice.

> 3. On the right side, under configurations, click on the in *Step 1 Startup / Bootstrap of the child device* created config file (`config1`) in the column: `DEVICE-SUPPORTED CONFIGURATIONS` click on the config file (`config1`) and click on `Get snapshot from device`. This will trigger a config snapshot requests via MQTT.

Reviewer: ```suggestion
3. On the right side, under configurations, click on the config file (`config1`) created in *Step 1 Startup / Bootstrap of the child device* in the column: `DEVICE-SUPPORTED CONFIGURATIONS` and then click on `Get snapshot from device`. This will trigger a config snapshot request to the child device via MQTT.
```
An image here with those buttons highlighted would be nice.

> 4. Send an “executing” operation status update to acknowledge the receipt of the request via MQTT. After receiving of the request, the agent must send an "executing" MQTT status message as follows:

Reviewer: ```suggestion
4. After receiving the request, the child device connector may **optionally** acknowledge the receipt of the request by sending an "executing" MQTT status message, as follows:
```

**Topic:**

`tedge/{child-d}/commands/res/config_snapshot`

**Payload**:

```json
{
    "status": "executing",
    "type": "{config-type}",
    "path": "/child/local/fs/path"
}
```

Example:

> `
> mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_snapshot" -m '{"status": "executing", "path": "/home/pi/config/config1", "type": "config1"}'
> `

Reviewer: ```suggestion
```console
mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_snapshot" -m '{"status": "executing", "path": "/home/pi/config/config1", "type": "config1"}'
```
```

5. Upload the requested config file to the URL received in the request via HTTP. After sending this 'executing' status message, the agent must upload the requested configuration file content to the url received in the request with an HTTP PUT request.

Example:

`
curl -X PUT --data-binary @/home/pi/config/config2 http://127.0.0.1:8000/tedge/file-transfer/child1/config_snapshot/config1
`

> 6. Send a “successful” operation status update via MQTT. Once the upload is complete, send a "successful" MQTT status message as follows:

Reviewer: ```suggestion
6. Once the upload is complete, send a "successful" status message via MQTT as follows:
```

**Topic:**

`tedge/{child-d}/commands/res/config_snapshot`

**Payload**:

```json
{
    "status": "successful",
    "type": "{config-type}",
    "path": "/child/local/fs/path"
}
```

Example:

```
mosquitto_pub -h localhost -t "tedge/child1/commands/res/config_snapshot" -m '{"status": "successful", "path": "/home/pi/config/config1", "type": "config1"}'
```

> ### Step 3 Handle config update request from thin-edge.io

Reviewer: ```suggestion
### Step 3 Handle config update request
```

> Handling config update requests from thin-edge.io is a 5-step process:

Reviewer: ```suggestion
Performing config update is an 8-step process:
```

> 1. Subscribe to, and receive config update requests via MQTT. In Cumulocity IoT, go to `device management` ->  `Management` -> `Configuration repository` -> `Add configuration snapshot` and fill in the fields in the pop-up . Upload a simple text file will be sufficient as an example.Give it the name config1.

Reviewer: ```suggestion
1. Subscribe to, and receive config update requests via MQTT. Example: `mosquitto_sub -h 127.0.0.1 -t "tedge/child1/commands/req/config_update"`
2. In Cumulocity IoT, go to `device management` ->  `Management` -> `Configuration repository` -> `Add configuration snapshot` and fill in the fields in the pop-up . Upload a simple text file will be sufficient as an example.Give it the name config1.
```

> 1. Subscribe to, and receive config update requests via MQTT. In Cumulocity IoT, go to `device management` ->  `Management` -> `Configuration repository` -> `Add configuration snapshot` and fill in the fields in the pop-up . Upload a simple text file will be sufficient as an example.Give it the name config1.

Reviewer: Show how the **request** JSON looks like, with the `url` and other fields.

Author: JSON is added

2. Click on `Devices` -> `All Devices`, then click on your device. Then click on  `Child devices` ->  then click on the in *Step 1 Startup / Bootstrap of the child device* defined child device (child1) and then click on `Configuration`.

> 3. On the right side, under configurations, click on the in *Step 1 Startup / Bootstrap of the child device* created config file (`config1`) in the column: `DEVICE-SUPPORTED CONFIGURATIONS` click on the config file (`config1`)

Reviewer: The same mistakes in the phrasing of these sentences, highlighted in the earlier "config snapshot" section. is repeated here as well. Please correct those.

> 4. In the column *AVAILABLE SUPPORTED CONFIGURATIONS* click on the previously added/uploaded configuration snapshot (config1) and click on *Send configuration to device*. This will trigger receiving config snapshot requests via MQTT

Reviewer: Provide Cumulocity UI images at least for the steps 1, 3 and 4.

Author: images are added as requested

The child device agent must subscribe to the `tedge/{child-d}/commands/req/config_update` MQTT topic
to receive the config update requests from thin-edge.io.
These requests arrive in the following JSON format:

> 5. Send an “executing” operation status update to acknowledge the receipt of the request via MQTT. On receipt of the request, the agent must send an "executing" MQTT status message as follows:

Reviewer: Redundant sentences.

Author: removed redundancy

**Topic:**

`tedge/{child-d}/commands/res/config_update`

**Payload**:

```json
{
    "status": "executing",
    "type": "{config-type}",
    "path": "/child/local/fs/path"
}
```

Example:

`

> mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_snapshot" -m '{"status": "executing", "path": "/home/pi/config/config1", "type": "config1"}'

Reviewer: ```suggestion
mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_update" -m '{"status": "executing", "path": "/home/pi/config/config1", "type": "config1"}'
```

`

> 6. Download the config file update from the URL received in the request via HTTP. After sending this status message, the agent must download the configuration file update

Reviewer: ```suggestion
6. Download the config file update from the URL received in the request via HTTP.

After sending the "executing" status message, the agent must download the configuration file update
```

from the `url` received in the request with an HTTP GET request.
The agent can then apply the downloaded configuration file update on the device.

Example:

`
curl http://127.0.0.1:8000/tedge/file-transfer/child1/config_update/config1 --output config1
`

7. Apply the config file update on the child device

The agent can then apply the downloaded configuration file update on the device.

8. Send a “successful” operation status update via MQTT

Once the update is applied, send a "successful" MQTT status message as follows:

**Topic:**

`tedge/{child-d}/commands/res/config_update`

**Payload**:

```json
{
    "status": "successful",
    "type": "{config-type}",
    "path": "/child/local/fs/path"
}
```
Example:

`

> mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_snapshot" -m '{"status": "successful", "path": "/home/pi/config/config1", "type": "config1"}'

Reviewer: ```suggestion
mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_updatet" -m '{"status": "successful", "path": "/home/pi/config/config1", "type": "config1"}'
```

`

> If there are any failures while downloading and applying the update, to remove the entries in Cumulocity IoT, a "failed" status update must be sent instead, to the same topic as follows:

Reviewer: ```suggestion
If there are any failures while downloading and applying the update, fail the operation in Cumulocity IoT by sending a "failed" status update with the `reason` to the same topic as follows:
```

```json
{
    "status": "failed",
    "type": "{config-type}",

>     "path": "/child/local/fs/path"

Reviewer: ```suggestion
    "path": "/child/local/fs/path",
    "reason": "Download failed"
```

}
```

Example:

`

> mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_snapshot" -m '{"status": "failed", "path": "/home/pi/config/config1", "type": "config1"}'

Reviewer: ```suggestion
mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_update" -m '{"status": "failed", "reason": "Download failed", "path": "/home/pi/config/config1", "type": "config1"}'
```

`

## References

> 

Reviewer: Provide a reference to the how-to as well.

Author: is added

`
http://{tedge-ip}:8000/tedge/file-transfer/{child-id}/c8y-configuration-plugin
`

* `{tedge-ip}` is the IP of the thin-edge.io device which is configured as `mqtt.external.bind_address` or `mqtt.bind_address` or `127.0.0.1` if neither is configured.
* `{child-id}` is the child-device-id

Example:
```console
curl -X PUT --data-binary @/home/pi/config/c8y-configuration-plugin.toml http://127.0.0.1:8000/tedge/file-transfer/child1/c8y-configuration-plugin
`

3. Notify thin-edge.io about the upload via MQTT

Once the upload is complete, the agent should notify thin-edge.io about the upload by sending the following MQTT message:

Topic:

`tedge/{child-d}/commands/res/config_snapshot`

Payload:

`{ "type": "c8y-configuration-plugin”, "path": ”/child/local/fs/path”}`

Example:

Note: For using the `mosquito-pub` command (and later on in this tutorial also the `mosquitto_sub` command), you need to install the  mosquitto clients via `sudo apt-get install mosquitto-clients`

`mosquitto_pub -h 127.0.0.1 -t "tedge/child1/commands/res/config_snapshot" -m '{"path": "", "type":"c8y-configuration-plugin"}'
`

> ### Step 2 Handle config snapshot request

Author: done
